File: src/forms/mod1.py
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By

from .form_base import form_base
from getpass import getpass

import sys
import re
import time

from forms.form_extract import extract_and_save_form_controls

class mod1(form_base):
    
    def __init__(self, _driver, _data, _mode):
        super(mod1, self).__init__(_driver, _data, _mode)
        self.manage_applications_url = None
        self.main_application_handle = None
        self.entry_url = 'https://sydneystudent.sydney.edu.au/sitsvision/wrd/siw_lgn'

    def create_profile(self):
        CreateProfileForm(self.driver, self.data).run()

    def login_session(self):
        """
        Logs in to the application and returns the main window handle
        """
        try:
            driver = self.driver
            
            # Set a max retry count
            max_retries = 3
            current_try = 0
            
            while current_try < max_retries:
                try:
                    driver.get(self.entry_url)
                    
                    # Ensure the browser is still open after navigation
                    if not driver.window_handles:
                        raise Exception("Browser window was closed")
                    
                    # Set explicit wait
                    wait = WebDriverWait(driver, 30)
                    
                    # Wait for page to load
                    wait.until(EC.presence_of_element_located((By.TAG_NAME, "body")))
                    
                    # Store the main window handle
                    main_window = driver.current_window_handle
                    # LoginForm(driver, self.data).run()

                    if not self.collect_mode:
                        print('\n\n ================ personal info ================= \n', self.data[-1][3])
                        print('\n\n ================= edu info: ================= \n', self.data[-1][1])
                        print('\n\n ================= application info: ================= \n', self.data[-1][2])
                    return main_window
                    
                except Exception as e:
                    current_try += 1
                    print(f"Login attempt {current_try} failed: {str(e)}")
                    
                    if current_try >= max_retries:
                        raise
                    
                    # If browser was closed, reinitialize it
                    if "no such window" in str(e).lower() or not driver.window_handles:
                        print("Browser window was closed. Reinitializing...")
                        self.initialize_driver()
                        driver = self.driver
                    
                    # Wait before retrying
                    time.sleep(2)
        
        except Exception as e:
            print(f"Fatal error in login_session: {str(e)}")
            # Ensure the browser is closed on error
            try:
                driver.quit()
            except:
                pass
            raise

        self.manage_applications_url = driver.current_url
        self.main_application_handle = driver.current_window_handle
        
        if not self.collect_mode:
            print('\n\n ================ personal info ================= \n', self.data[-1][3])
            print('\n\n ================= edu info: ================= \n', self.data[-1][1])
            print('\n\n ================= application info: ================= \n', self.data[-1][2])

        return self.main_application_handle

    def fill_form(self):
        driver = self.driver
        
        return True

    def new_application(self):
        driver = self.driver
        students = self.data

        driver.close()
        driver.switch_to.window(self.main_application_handle)
        driver.get(self.manage_applications_url)
        students.pop()
        if not len(students):
            print('Congrats, you finished processing. ')
            return False

        print('Now processing: ', students[-1][0])
        return True

    def search_course(self):
        students = self.data

        df_application = students[-1][2]
        df_application = df_application[df_application['Proposed School'] == 'USYD']
        course_applied = df_application['Proposed Course with Corresponding Links'].tolist()[0]
        if df_application.shape[0] > 1:
            df_application = df_application.drop(index=[0])
            new_item = students[-1].copy()
            new_item[0]['Number'] += 1
            new_item[2] = df_application
            students.insert(0, new_item)
            
        course = '//*[contains(@id,"POP_UDEF") and contains(@id,"POP.MENSYS.1-1")]'
        self.set_value(course, course_applied)

        return

    def payment_email(self):
        email = '//*[@id="UDS_EMAIL"]'
        self.set_value(email, 'au.info@shinyway.com')

    def payment_method(self):
        #cardnumber = '//*[@id="UDS_CARDNUMBER"]'
        cardnumber = '/html/body/form/fieldset/ul/li[1]/div[2]/input'
        month = '/html/body/form/fieldset/ul/li[2]/div[2]/input[1]'
        year = '/html/body/form/fieldset/ul/li[2]/div[2]/input[2]'
        self.set_value(cardnumber, '123456789')
        self.set_value(month, '09')
        self.set_value(year, '25')

    def run(self):
        if self.collect_mode:
            print('collect page information.')
            return

        students = self.data
        if not len(students):
            print('no more studnets to process.')
            sys.exit()
        
        wait = WebDriverWait(self.driver, 10)
        wait.until(EC.presence_of_element_located((By.TAG_NAME, "body")))
        url = self.driver.current_url

        if re.search('https://sydneystudent.sydney.edu.au/sitsvision/wrd/siw_ipp_cgi.start?', url):
            if not self.fill_form():
                sys.exit()

        elif re.search('https://sydneystudent.sydney.edu.au/sitsvision/wrd/SIW_POD.start_url?.+', url):
            self.search_course()

        # create profile
        elif re.search('https://sydneystudent.sydney.edu.au/sitsvision/wrd/SIW_YMHD.start_url?.+', url):
            self.create_profile()

        # email payment
        elif re.search('https://pay.sydney.edu.au/ePays/paymentemail\?UDS_ACTION=PMPBPN', url):
            self.payment_email()
            
        elif re.search('https://pay.sydney.edu.au/ePays/paymentemail$', url):
            self.payment_method()

        else:
            print('no actions for: ', url)
            pass

File: browser-extension/forms/mod1.js
const { Key, By, until } = require('selenium-webdriver');
const FormBase = require('./form_base');
const { extractAndSaveFormControls } = require('./form_extract');
const readline = require('readline');

class Mod1 extends FormBase {
  constructor(_driver, _data, _mode) {
    super(_driver, _data, _mode);
    this.manageApplicationsUrl = null;
    this.mainApplicationHandle = null;
    this.entryUrl = 'https://sydneystudent.sydney.edu.au/sitsvision/wrd/siw_lgn';
  }

  async createProfile() {
    // Implementation for CreateProfileForm would go here
    // This is a placeholder - we would need the actual CreateProfileForm class
    console.log("Creating profile...");
  }

  async loginSession() {
    /**
     * Logs in to the application and returns the main window handle
     */
    try {
      const driver = this.driver;
      
      // Set a max retry count
      const maxRetries = 3;
      let currentTry = 0;
      
      while (currentTry < maxRetries) {
        try {
          await driver.get(this.entryUrl);
          
          // Ensure the browser is still open after navigation
          const handles = await driver.getAllWindowHandles();
          if (handles.length === 0) {
            throw new Error("Browser window was closed");
          }
          
          // Set explicit wait
          const wait = driver.wait(until.elementLocated(By.tagName("body")), 30000);
          
          // Wait for page to load
          await wait;
          
          // Store the main window handle
          const mainWindow = await driver.getWindowHandle();
          // LoginForm would be implemented here if needed
          
          if (!this.collectMode) {
            console.log('\n\n ================ personal info ================= \n', this.data[this.data.length - 1][3]);
            console.log('\n\n ================= edu info: ================= \n', this.data[this.data.length - 1][1]);
            console.log('\n\n ================= application info: ================= \n', this.data[this.data.length - 1][2]);
          }
          return mainWindow;
          
        } catch (e) {
          currentTry++;
          console.log(`Login attempt ${currentTry} failed: ${e.toString()}`);
          
          if (currentTry >= maxRetries) {
            throw e;
          }
          
          // If browser was closed, reinitialize it
          if (e.toString().toLowerCase().includes("no such window") || !(await driver.getAllWindowHandles()).length) {
            console.log("Browser window was closed. Reinitializing...");
            // Reinitialize driver would happen here
            // this.initializeDriver();
            driver = this.driver;
          }
          
          // Wait before retrying
          await new Promise(resolve => setTimeout(resolve, 2000));
        }
      }
    } catch (e) {
      console.log(`Fatal error in login_session: ${e.toString()}`);
      // Ensure the browser is closed on error
      try {
        await this.driver.quit();
      } catch {
        // Ignore errors during quit
      }
      throw e;
    }

    this.manageApplicationsUrl = await this.driver.getCurrentUrl();
    this.mainApplicationHandle = await this.driver.getWindowHandle();
    
    if (!this.collectMode) {
      console.log('\n\n ================ personal info ================= \n', this.data[this.data.length - 1][3]);
      console.log('\n\n ================= edu info: ================= \n', this.data[this.data.length - 1][1]);
      console.log('\n\n ================= application info: ================= \n', this.data[this.data.length - 1][2]);
    }

    return this.mainApplicationHandle;
  }

  async fillForm() {
    const driver = this.driver;
    
    return true;
  }

  async newApplication() {
    const driver = this.driver;
    const students = this.data;

    await driver.close();
    await driver.switchTo().window(this.mainApplicationHandle);
    await driver.get(this.manageApplicationsUrl);
    students.pop();
    if (students.length === 0) {
      console.log('Congrats, you finished processing.');
      return false;
    }

    console.log('Now processing: ', students[students.length - 1][0]);
    return true;
  }

  async searchCourse() {
    const students = this.data;

    const dfApplication = students[students.length - 1][2];
    const filteredDf = dfApplication.filter(row => row['Proposed School'] === 'USYD');
    const courseApplied = filteredDf[0]['Proposed Course with Corresponding Links'];
    if (filteredDf.length > 1) {
      const newDf = filteredDf.slice(1);
      const newItem = [...students[students.length - 1]];
      newItem[0]['Number'] += 1;
      newItem[2] = newDf;
      students.unshift(newItem);
    }
            
    const course = '//*[contains(@id,"POP_UDEF") and contains(@id,"POP.MENSYS.1-1")]';
    await this.setValue(course, courseApplied);
  }

  async paymentEmail() {
    const email = '//*[@id="UDS_EMAIL"]';
    await this.setValue(email, 'au.info@shinyway.com');
  }

  async paymentMethod() {
    const cardnumber = '/html/body/form/fieldset/ul/li[1]/div[2]/input';
    const month = '/html/body/form/fieldset/ul/li[2]/div[2]/input[1]';
    const year = '/html/body/form/fieldset/ul/li[2]/div[2]/input[2]';
    
    await this.setValue(cardnumber, '123456789');
    await this.setValue(month, '09');
    await this.setValue(year, '25');
  }

  async run() {
    if (this.collectMode) {
      console.log('collect page information.');
      return;
    }

    const students = this.data;
    if (students.length === 0) {
      console.log('no more students to process.');
      process.exit();
    }
    
    const wait = this.driver.wait(until.elementLocated(By.tagName("body")), 10000);
    await wait;
    const url = await this.driver.getCurrentUrl();

    if (url.match('https://sydneystudent.sydney.edu.au/sitsvision/wrd/siw_ipp_cgi.start?')) {
      if (!await this.fillForm()) {
        process.exit();
      }
    } else if (url.match('https://sydneystudent.sydney.edu.au/sitsvision/wrd/SIW_POD.start_url?.+')) {
      await this.searchCourse();
    } else if (url.match('https://sydneystudent.sydney.edu.au/sitsvision/wrd/SIW_YMHD.start_url?.+')) {
      await this.createProfile();
    } else if (url.match('https://pay.sydney.edu.au/ePays/paymentemail\\?UDS_ACTION=PMPBPN')) {
      await this.paymentEmail();
    } else if (url.match('https://pay.sydney.edu.au/ePays/paymentemail$')) {
      await this.paymentMethod();
    } else {
      console.log('no actions for: ', url);
    }
  }
}

module.exports = Mod1;
